src/uharfbuzz/_harfbuzz.pyx

```diff
@@ -1057,3 +1057,152 @@ cdef class DrawFuncs:
+        return NULL
+    else:
+        p = <hb_link_t*>calloc(link_num, sizeof(hb_link_t))
+        if p == NULL:
+            free_allocated(obj_list, length)
+            # for virtual_links, need to free additional real_links which
+            # is created before virtual_links
+            if obj_list[length].real_links != NULL:
+                free(obj_list[length].real_links)
+            free(obj_list)
+
+            success[0] = 0
+            return NULL
+        else:
+            return p
+
+cdef void free_allocated(hb_object_t* obj_list, unsigned int length):
+    cdef unsigned int i
+    for i in range(length):
+        if obj_list[i].real_links != NULL:
+            free(obj_list[i].real_links)
+        if obj_list[i].virtual_links != NULL:
+            free(obj_list[i].virtual_links)
+
+# The whole table is represented as a Graph
+# The input graphnodes is a flat list of subtables
+# each Node(subtable) is represented by a tuple of 2 lists: real_links and virtual_links
+#
+# real_links is a list of edges for this node (offset field in the table)
+# An real edge is represented by a tuple of (int position, int width, int
+# objidx):
+#
+# - position: means relative position of the offset field in bytes from the 
+#             beginning of the subtable's C struct. e.g:
+#   a GSUB header struct in C looks like below:
+#   uint16 majorVersion
+#   uint16 minorVersion
+#   offset16 scriptListOffset
+#   offset16 featureListOffset
+#   offset16 lookupListOffset
+#   And the position for scriptListOffset is 4 which is calculated from (16+16)/8
+# - width: size of the offset (e.g 2 for offset16 and 4 for offset32)
+# - objidx: objidx is the index of the subtable in graph/tree generated by
+#           postorder traversal
+#
+# virtual_links is a list of virtual edges, they are not real offset links,
+# a virtual_link specifies an ordering constraint that harfbuzz packing must follow.
+# A virtual edge would have 0 offsetSize, 0 position, and a real objidx.
+# e.g, if Node A has a virtual edge with objidx b(corresponding Node is B),
+# then that means Node B is always packed after Node A in the final serialized order
+def hb_repack(subtables: List[bytes],
+              graphnodes: List[Tuple[List[Tuple[int, int, int]],
+                                     List[Tuple[int, int, int]]
+                                     ]]) -> bytes:
+    if len(subtables) != len(graphnodes):
+        return bytes(b'')
+    cdef:
+        unsigned int num_nodes = len(subtables)
+        unsigned int length = 0, prev_length = 0
+        dict idx_len_dict = {}
+    for i in range(num_nodes):
+        length += len(subtables[i])
+        idx_len_dict[i] = (prev_length, length)
+        prev_length = length
+
+    cdef bytes table_bytes = b''.join(subtables)
+    cdef char* table_data = table_bytes
+
+    cdef hb_object_t* obj_list = <hb_object_t*>calloc(num_nodes, sizeof(hb_object_t))
+    if obj_list == NULL:
+      return bytes(b'')
```

Comment:
Shouldn’t this be a more obvious error than just returning an empty string? Same for the other cases below.